src/test/common/clara/test_memory.cljc

```diff
@@ -43,10 +44,48 @@
 
 (use-fixtures :once st/validate-schemas #?(:clj tu/opts-fixture))
 
+(defn join-filter-equals
+  "Intended to be a test function that is the same as equals, but is not visible to Clara as such
+  and thus forces usage of join filters instead of hash joins"
+  [& args]
+  (apply = args))
+
 ;; While the memory is tested through rules and queries, rather than direct unit tests on the memory,
 ;; the intent of these tests is to create patterns in the engine that cover edge cases and other paths
 ;; of concern in clara.rules.memory.
 
+(def-rules-test test-negation-complex-join-with-numerous-non-matching-facts-inserted-after-descendant-negation
+  {:rules []
+   :queries [query1 [[]
+                     [[Hot (= ?t temperature)]
+                      [:not [Cold (join-filter-equals ?t temperature)]]]]]
+   :sessions [empty-session [query1] {}]}
+  (let [lots-of-hot (doall (for [_ (range 100)]
+                             (->Hot 20)))]
+    (is (= (repeat 100 {:?t 20})
+           (-> empty-session
+               (insert (->Cold 10))
+               fire-rules
+               (insert-all lots-of-hot)
+               fire-rules
+               (query query1))))))
+
+(def-rules-test test-negation-complex-join-with-numerous-non-matching-facts-inserted-after-descendant-negation-with-equals
```

Comment:
Sort of unfortunate the whole test body has to be repeated just to change `join-filter-equals` to `=`.
Maybe that is an artifact of `def-rules-test` and perhaps it could be improved later.

Maybe a comment here saying this is the same as above though would be good to make it clear they are supposed to be in sync.